lib/api/objectCopy.js

```diff
@@ -91,15 +93,31 @@ function _prepMetadata(sourceObjMD, headers, sourceIsDestination, authInfo,
         });
         return { error: userMetadata };
     }
-    // If metadataDirective is 'COPY' and location constraint header is
-    // specified, new location constraint should override source object's
-    if (whichMetadata === 'COPY' && headers[locationHeader]) {
-        userMetadata[locationHeader] = headers[locationHeader];
+    // If metadataDirective is:
+    // - 'COPY' and source object has a locaction constrain in its metadata
+    // we use the bucket destination location constraint
+    if (whichMetadata === 'COPY'
+        && userMetadata[locationHeader]
+        && destBucketMD.getLocationConstraint()) {
+        userMetadata[locationHeader] = destBucketMD.getLocationConstraint();
```

Comment:
Based on this, are we ignoring object level location constraint which may have been on the copy object request, even if it takes priority over bucket location constraint?

And just to confirm what we do when REPLACE:
If copy object request has obj-lvl location constraint, we copy to dest location header?
If copy object request has bucket-lvl location constraint and no obj-lvl location constraint, and source object had a location header, we copy the bucket-lvl location constraint to the dest location header?

Reply:
Here we use the metadata header from the request if copy https://github.com/scality/S3/pull/942/files#diff-5ce9bf7b97556522975c2fae1a6944efR86

Yes, If copy object request has `x-amz-meta-scal-location-constraint`, we copy to dest location constraint

If you copy an object with `azure-test` location using `REPLACE` and no `x-amz-meta-scal-location-constraint` to a bucket with `mem-test` location constraint , the copied object location constraint will be empty so we will use the location of the destination bucket, ie `mem-test`.

Does it make sense?

Reply:
At line 86, `userMetadata` looks like the source object MD, not the request headers (if COPY). ?

Can you give me the line # for where you copy the request `x-amz-meta-scal-location-constraint` to the dest location constraint? I think it will be quicker for you to find than for me to look.

> If you copy an object with azure-test location using REPLACE and no x-amz-meta-scal-location-constraint to a bucket with mem-test location constraint , the copied object location constraint will be empty so we will use the location of the destination bucket, ie mem-test.
It makes sense but where do you copy the destination bucket location constraint? Since in the condition above it looks like it only applies for `COPY`?

Reply:
Sorry I mean "Here we use the metadata header from the request if **REPLACE**" https://github.com/scality/S3/pull/942/files#diff-5ce9bf7b97556522975c2fae1a6944efR88

Reply:
OK, as discussed:
- when `REPLACE` and there was source object `x-amz-meta-scal-location-constraint` but no request `x-amz-meta-scal-location-constraint`, we leave dest `x-amz-meta-scal-location-constraint` empty and default to using the bucket location constraint
- when `COPY` we disregard the request headers for metadata, since in concept we are pulling the metadata from the original object.

Thanks for the clarification!